File: sorting/MergeSort/python/MergeSort.py
from __future__ import annotations


def merge_sort(values: list[int]) -> list[int]:
    """Return a stable, nondecreasing ordering of the input values."""
    sorted_values = list(values)
    buffer = [0] * len(sorted_values)

    def sort_range(left: int, right: int) -> None:
        if right - left <= 1:
            return

        mid = left + (right - left) // 2
        sort_range(left, mid)
        sort_range(mid, right)

        i = left
        j = mid
        k = left

        while i < mid and j < right:
            if sorted_values[i] <= sorted_values[j]:
                buffer[k] = sorted_values[i]
                i += 1
            else:
                buffer[k] = sorted_values[j]
                j += 1
            k += 1

        while i < mid:
            buffer[k] = sorted_values[i]
            i += 1
            k += 1

        while j < right:
            buffer[k] = sorted_values[j]
            j += 1
            k += 1

        sorted_values[left:right] = buffer[left:right]

    sort_range(0, len(sorted_values))
    return sorted_values

File: sorting/MergeSort/cpp/MergeSort.cpp
#include "MergeSort.hpp"

#include <algorithm>

using namespace std;

namespace {
void mergeRanges(vector<int64_t>& values, vector<int64_t>& buffer, size_t left, size_t mid, size_t right) {
    size_t i = left;
    size_t j = mid;
    size_t k = left;

    while (i < mid && j < right) {
        if (values[i] <= values[j]) {
            buffer[k++] = values[i++];
        } else {
            buffer[k++] = values[j++];
        }
    }

    while (i < mid) {
        buffer[k++] = values[i++];
    }

    while (j < right) {
        buffer[k++] = values[j++];
    }

    copy(buffer.begin() + static_cast<long long>(left), buffer.begin() + static_cast<long long>(right), values.begin() + static_cast<long long>(left));
}

void mergeSortRecursive(vector<int64_t>& values, vector<int64_t>& buffer, size_t left, size_t right) {
    if (right - left <= 1) {
        return;
    }

    const size_t mid = left + (right - left) / 2;
    mergeSortRecursive(values, buffer, left, mid);
    mergeSortRecursive(values, buffer, mid, right);
    mergeRanges(values, buffer, left, mid, right);
}
}  // namespace

vector<int64_t> mergeSort(const vector<int64_t>& values) {
    vector<int64_t> sorted = values;
    vector<int64_t> buffer(sorted.size());
    mergeSortRecursive(sorted, buffer, 0, sorted.size());
    return sorted;
}
